lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,849 @@
+        PR_ASSERT(PR_FALSE);
+        return CKR_MECHANISM_PARAM_INVALID;
+    }
+
+    return CKR_OK;
+}
+
+static CK_RV kbkdf_AddParameters(CK_MECHANISM_TYPE mech, SftkMACCtx *ctx, CK_SP800_108_KDF_PARAMS *params, uint32_t counter, uint64_t length, uint8_t *chaining_prf) {
+    size_t offset = 0;
+    CK_RV ret;
+
+    for (offset = 0; offset < params->ulNumberOfDataParams; offset++) {
+        CK_PRF_DATA_PARAM_PTR param = params->pDataParams + offset;
+
+        switch (param->type) {
+        case CK_SP800_108_ITERATION_VARIABLE:
+            /* When present in COUNTER mode, this signifies adding the counter
+             * variable to the PRF. Otherwise, it signifies the chaining
+             * value for other KDF modes. */
+            if (mech == CKM_SP800_108_COUNTER_KDF) {
+                CK_SP800_108_COUNTER_FORMAT_PTR counter_format = (CK_SP800_108_COUNTER_FORMAT_PTR) param->pValue;
+                CK_BYTE buffer[8];
+                CK_ULONG num_bytes;
+                kbkdf_EncodeInteger(counter, counter_format->ulWidthInBits, counter_format->bLittleEndian, buffer, &num_bytes);
+                ret = sftk_MAC_Update(ctx, buffer, num_bytes);
+            } else {
+                /* It is safe to pass chaining_ptr when the KDF is in counter
+                 * mode. */
+                PR_ASSERT(chaining_prf != NULL);
+                ret = sftk_MAC_Update(ctx, chaining_prf, ctx->output_size);
+            }
+            break;
+        case CK_SP800_108_COUNTER:
+            /* Only present in the case when not using COUNTER mode. */
+            PR_ASSERT(mech != CKM_SP800_108_COUNTER_KDF);
+            break;
+        case CK_SP800_108_BYTE_ARRAY:
+            ret = sftk_MAC_Update(ctx, (CK_BYTE_PTR) param->pValue, param->ulValueLen);
+            break;
+        case CK_SP800_108_DKM_LENGTH: {
+            /* We've already done the hard work of calculating the length in
+             * the kbkdf_CalculateIterations function; we merely need to add
+             * the length to the desired point in the input stream. */
+            CK_SP800_108_DKM_LENGTH_FORMAT_PTR length_format = (CK_SP800_108_DKM_LENGTH_FORMAT_PTR) param->pValue;
+            CK_BYTE buffer[8];
+            CK_ULONG num_bytes;
+            kbkdf_EncodeInteger(length, length_format->ulWidthInBits, length_format->bLittleEndian, buffer, &num_bytes);
+            ret = sftk_MAC_Update(ctx, buffer, num_bytes);
+            break;
+                                      }
+        default:
+            PR_ASSERT(PR_FALSE);
+            break;
+        }
+
+        if (ret != CKR_OK) {
+            PR_ASSERT(PR_FALSE);
+            return ret;
+        }
+    }
+
+    return CKR_OK;
+}
+
+CK_RV kbkdf_SaveKey(SFTKObject *key, uint8_t *key_buffer, size_t key_len) {
+    /* This is at least necessary under kbkdf_CreateKey; might as well always
+     * do it. TODO: Figure out if there's any harm in always doing it. */
```

Comment:
This should be fine. The mechanism always derives CKO_SECRET_KEYs. (as opposed to CKO_PRIVATE_KEYS or CKO_PUBLIC_KEYS. The one exception is the extract case, which is CKO_DATA_OBJECT, but in that case we could adjust the result before we return it.

Reply:
This section will be rewritten, sorry -- I'm working on making the KBKDF interface match the MAC interface that I proposed above -- a `kbkdf_Counter(.... SFTKObject *key ... )` and a `kbkdf_CounterRaw(.... uint8_t *key, uint32_t key_length, ... )` -- that'll make the TLS implementation easier (no need to handle creating a fake key, passing it to the KDF, and then unwrapping and freeing it).

Then if we want a CKO_DATA_OBJECT, we can call Raw instead and I think it'll be cleaner .